### PR TITLE
Update sdk to version 25 and fix ResourceClassGenerator to not write final fields

### PR DESCRIPTION
**Pull request by hameno**

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -43,8 +43,8 @@
     <asm.version>5.0.4</asm.version>
     <!-- Attention - make sure that the tools version is the versions of 
       the transitive dependencies of the builder version -->
-    <android.builder.version>1.5.0</android.builder.version>
-    <android.tools.version>24.5.0</android.tools.version>
+    <android.builder.version>2.2.3</android.builder.version>
+    <android.tools.version>25.2.0</android.tools.version>
```
Comment (Shusshu) [inline pom.xml]: 25.2.3  and if we have too many problems we can try the latest beta

Comment (hameno) [inline pom.xml]: > Sorry this is only on jcenter.....
Argh, hate what has happened to the maven ecosystem lately...

File: pom.xml [hunk 2]
```diff
@@ -43,8 +43,8 @@
     <asm.version>5.0.4</asm.version>
     <!-- Attention - make sure that the tools version is the versions of 
       the transitive dependencies of the builder version -->
-    <android.builder.version>1.5.0</android.builder.version>
-    <android.tools.version>24.5.0</android.tools.version>
+    <android.builder.version>2.2.3</android.builder.version>
+    <android.tools.version>25.2.3</android.tools.version>
   </properties>
```
Comment (Shusshu) [inline pom.xml]: you need this too:

  <repositories>
    <repository>
      <snapshots>
        <enabled>false</enabled>
      </snapshots>
      <id>bintray-maven</id>
      <name>bintray</name>
      <url>https://dl.bintray.com/android/android-tools</url>
    </repository>
  </repositories>

**Comment by Shusshu** [comment]
I'm updating stuff on my branch https://github.com/simpligility/android-maven-plugin/tree/update-sdk-25

**Comment by Shusshu** [comment]
Now I remember, last time I couldn't find how to fix that change but you did it :)

 final String packageName = VariantConfiguration.getManifestPackage( libManifestFile );
-> final String packageName = new DefaultManifestParser( libManifestFile ).getPackage();

**Comment by hameno** [comment]
Yeah, I found that code in the old version, ```VariantConfiguration.getManifestPackage``` basically just called the ```DefaultManifestParser```...

**Comment by Shusshu** [comment]
So in my branch only the emulator test is not working... and the manifest merger 2 needs to be checked

**Comment by Shusshu** [comment]
Btw some code is gone from the latest beta ...
```
import com.android.builder.internal.SymbolLoader;
import com.android.builder.internal.SymbolWriter; 
```

**Comment by Shusshu** [comment]
I've updated my branch with the latest BETA stuff

**Comment by Shusshu** [comment]
SymbolWiter is deprecated https://android.googlesource.com/platform/tools/base/+/master/sdklib/src/main/java/com/android/sdklib/internal/build/SymbolWriter.java

They say to use AndroidBuilder https://android.googlesource.com/platform/tools/build/+/master/builder/src/main/java/com/android/builder/AndroidBuilder.java 

And then I found 

package com.android.builder.symbols;

import com.android.builder.symbols.SymbolIo;
import com.android.builder.symbols.SymbolTable;
import com.android.utils.Pair;
import com.google.common.base.Preconditions;
import java.io.File;
import java.util.Collection;
import java.util.HashMap;
import java.util.HashSet;
import java.util.Iterator;
import java.util.function.Consumer;

```
public class RGeneration {
    private RGeneration() {
    }

    public static void generateRForLibraries(SymbolTable main, Collection<SymbolTable> libraries, File out, boolean finalIds) {
        Preconditions.checkArgument(out.isDirectory(), "!out.iDirectory");
        Pair mainP = Pair.of(main.getTablePackage(), main.getTableName());
        HashMap toWrite = new HashMap();
        Iterator var6 = libraries.iterator();

        while(var6.hasNext()) {
            SymbolTable k = (SymbolTable)var6.next();
            Pair st = Pair.of(k.getTablePackage(), k.getTableName());
            if(!st.equals(mainP)) {
                SymbolTable existing = (SymbolTable)toWrite.get(st);
                if(existing != null) {
                    toWrite.put(st, existing.merge(k));
                } else {
                    toWrite.put(st, k);
                }
            }
        }

        var6 = (new HashSet(toWrite.keySet())).iterator();

        while(var6.hasNext()) {
            Pair k1 = (Pair)var6.next();
            SymbolTable st1 = (SymbolTable)toWrite.get(k1);
            st1 = main.filter(st1).rename(st1.getTablePackage(), st1.getTableName());
            toWrite.put(k1, st1);
        }

        toWrite.values().forEach((st) -> {
            SymbolIo.exportToJava(st, out, finalIds);
        });
    }
}
```

**Comment by Shusshu** [comment]
I will let you finish this PR and properly test it :p

**Comment by hameno** [comment]
I'll try, any hint on my previous question ("I'm not sure how I can detect if the generation is run during test lifecycle (and inside an aar library).")

**Comment by Shusshu** [comment]
Maybe @mosabua knows

**Comment by hameno** [comment]
I've made some progress updating it to match AndroidBuilder#processResources, currently testing

**Comment by hameno** [comment]
Is there a IntelliJ codestyle config anywhere? Checkstyle kills me here...

**Comment by Shusshu** [comment]
haha I also hate the codestyle :p @mosabua Do we have an intellij formatter?

**Comment by hameno** [comment]
Huh, so now it only generates the main R file (but includes all symbols from libs), not yet sure how to fix that using the new classes.

**Comment by hameno** [comment]
I think that's it, my (rather complex 20-modules) project is now building successfully and all tests are working, even with robolectric 3.2.1. 
Now I need to verify, if all (manifests etc) is correct :)

Also: I still need to check an actual apk build (my project only consists of aar libraries)

**Comment by Shusshu** [comment]
Do you use manifest merger? Cuz I quickly did that upgrade and wasn't too
sure about that. I didn't take the time to look deeper.

It needs to be tested at least with one merge

**Comment by hameno** [comment]
Yes, it's used heavily in this project. It's on my to-do list for apk build test. Gonna compare old and new APK using Android studio.

**Comment by Shusshu** [comment]
Oh and manifest merger v1 is gone. Some class are gone from android
builder... Maybe we can bring them back in the project if we still want to
support it

**Comment by hameno** [comment]
APK looks good, will cleanup this branch and try to match the codestyle and then it should be done

**Comment by mosabua** [comment]
You rock guys! I cant keep up with you. If this ends up all working I will try to pull it in and cut a release asap.

**Comment by hameno** [comment]
Okay, I'm done. Not sure why those tests fail...I did update the jdk to java8 (I think this is due to the android tools requiring java8 now).

**Comment by Shusshu** [comment]
@simpligility/android-maven-plugins-core-committers  anyone wants to try this out?

File: pom.xml [hunk 3]
```diff
@@ -36,15 +36,15 @@
...
     <easymock.version>3.4</easymock.version>
     <powermock.version>1.6.4</powermock.version>
     <publisher.version>v2-rev13-1.19.0</publisher.version>
     <asm.version>5.0.4</asm.version>
     <!-- Attention - make sure that the tools version is the versions of 
       the transitive dependencies of the builder version -->
-    <android.builder.version>1.5.0</android.builder.version>
-    <android.tools.version>24.5.0</android.tools.version>
+    <android.builder.version>2.3.0-beta2</android.builder.version>
+    <android.tools.version>25.3.0-beta2</android.tools.version>
```
Comment (Shusshu) [inline pom.xml]: Change to 25.3.0-beta3

File: pom.xml [hunk 4]
```diff
@@ -36,15 +36,15 @@
...
+    <takari.test.version>2.9.1</takari.test.version>
     <easymock.version>3.4</easymock.version>
     <powermock.version>1.6.4</powermock.version>
     <publisher.version>v2-rev13-1.19.0</publisher.version>
     <asm.version>5.0.4</asm.version>
     <!-- Attention - make sure that the tools version is the versions of 
       the transitive dependencies of the builder version -->
-    <android.builder.version>1.5.0</android.builder.version>
-    <android.tools.version>24.5.0</android.tools.version>
+    <android.builder.version>2.3.0-beta2</android.builder.version>
```
Comment (Shusshu) [inline pom.xml]: change to 2.3.0-beta3

**Comment by mosabua** [comment]
I can just try it with the projects from the plugin.. hopefully in the next days. But if you are happy with it @Shusshu .. feel free to merge to master

**Comment by Shusshu** [comment]
@william-ferguson-au Happy with this?

**Comment by william-ferguson-au** [comment]
I get a bunch of tests failing:

```
Results :

Tests in error:
  AbstractEmulatorMojoTest.testStartAndroidEmulatorAlreadyBooted:93 » NoClassDefFound
  AbstractEmulatorMojoTest.testStartAndroidEmulatorWithBootTimeout:125 » NoClassDefFound
  AbstractEmulatorMojoTest.testStartAndroidEmulatorWithOngoingBoot:108 » NoClassDefFound
  AbstractEmulatorMojoTest.testStartAndroidEmulatorWithTimeoutToConnect:74 » ExceptionInInitializer
  AndroidSdkTest.validPlatformsAndApiLevels25:96 » InvalidSdk Invalid SDK: Build...
  AndroidSdkTest.validPlatformsAndApiLevelsWithDiffBuildTools2:129 » InvalidSdk ...
  AndroidSdkTest.validPlatformsAndApiLevelsWithDiffBuildToolsMinor:136 » InvalidSdk

Tests run: 96, Failures: 0, Errors: 7, Skipped: 11
```

**Comment by william-ferguson-au** [comment]
The AbstractEmulatorMojoTests are failing because it's not finding 'com.android.sdklib.repository.AndroidSdkHandler'. Where does that comes from?

The other tests were failing because I was missing build-tools 24.0.1 and 24.0.2

**Comment by Shusshu** [comment]
AndroidSdkHandler is used in the AndroidSdk class for loading the sdk

I see in the EmulatorMojoTest this:

        @Override
        protected AndroidSdk getAndroidSdk()
        {
            return new SdkTestSupport().getSdk_with_platform_default();
        }

Which in turns needs 

     sdk_with_platform_default = new AndroidSdk(new File(env_ANDROID_HOME), "19");

**Comment by Shusshu** [comment]
@hameno you can merge my last commit.

@simpligility/android-maven-plugins-core-committers It's either we merge and get the latest gradle android plugin in with a possible buggy emulator mojo (test fails but I actually never used that mojo)

**Comment by hameno** [comment]
@Shusshu Done

**Comment by william-ferguson-au** [comment]
I have never used the emulator mojo either.
Would love to get some metrics over which mojos are used and how much.

**Comment by Shusshu** [comment]
@hameno you can use rc1 now

    <android.builder.version>2.3.0-rc1</android.builder.version>
    <android.tools.version>25.3.0-rc1</android.tools.version>

**Comment by Shusshu** [comment]
You can update it again ;) it's released 2.3.0  and then somebody should test & merge this

**Comment by hameno** [comment]
I got every test except AbstractEmulatorMojoTest to pass.

AbstractEmulatorMojoTest fails due to an ExceptionInInitializerError because AndroidSdkHandler cannot init the SchemaModule fields (ADDON_MODULE, REPOSITORY_MODULE, SYS_IMG_MODULE, COMMON_MODULE). I tracked it down to SchemaModule:156 which tries to get a namespace for a XML annotation. I'm not sure why this fails as the LinkedList contains the annotation but with a different hash?! Maybe PowerMock is the reason? Should we just disable the test?

**Comment by Shusshu** [comment]
I don't mind if we disable it...  wdyt @simpligility/android-maven-plugins-core-committers ?

**Comment by mosabua** [comment]
If we disable support for the emulator mojos we completely loose the ability to run automated integration tests. That sucks imho. I am however not aware of the plans from the SDK team on how to support emulator interaction going forward (or even the current status).

**Comment by mosabua** [comment]
But yes.. maybe we can disable the test for now or at least the failing method and try to replace it. That whole mocking is a PITA anyway. If the mojo works in an integration test maybe that test can completely disappear and we dont worry about it.

**Comment by Shusshu** [comment]
@hameno can you add a @Ignore for the emulator test?

@simpligility/android-maven-plugins-core-committers I have been testing the emulator-start command and it actually doesn't work even in previous versions

The script it creates can be executed manually but the CommandExecutor cannot run it. I couldn't find why.

So let's merge this in with @Ignore on the emulator test

**Comment by hameno** [comment]
Okay, gonna do this later today. 

@Shusshu Maybe the script does not have execute permissions?

**Comment by Shusshu** [comment]
I was able to run it via the cmd line

**Comment by mosabua** [comment]
It sucks that the plugin needs the bintray repo. That basically means it wont work out of the box for normal Maven users and we need to update the docs. But I think we should merge this now and iterate in master..

**Comment by Shusshu** [comment]
Why would it not work for normal users?
It's just a repo... But anyway that was to access the android builder beta
version now we use the stable one so it should work without it

**Comment by mosabua** [comment]
Great ... you wanna merge this now then?

**Comment by mosabua** [comment]
I am working atm but hopefully can give this a whirl tonight.. just got back last night..

**Comment by Shusshu** [comment]
Ok cool. Just need to wait the last ignore change or you can do it later
when you try it out

**Comment by mosabua** [comment]
Sounds good. Will probably do a release associated to that .. have you tried the site build? With the change to JDK 8 it probably fails all over the place..

**Comment by Shusshu** [comment]
Nope haven't tried

**Comment by hameno** [comment]
I've disable the test

**Comment by mosabua** [comment]
Btw I updated the changedlog and remove the bintray repo. All good so far.. more tonight. Will up version to 4.5.0 for release

**Pull request merged**